cirq/circuits/circuit.py

```diff
@@ -1258,6 +1258,100 @@ def _json_dict_(self):
     def _from_json_dict_(cls, moments, device, **kwargs):
         return cls(moments, strategy=InsertStrategy.EARLIEST, device=device)
 
+    def raggedy_add(
+        *circuits: 'cirq.AbstractCircuit', stop_at_first_alignment: bool = False
+    ) -> 'cirq.Circuit':
+        """Concatenates circuits while overlapping them if possible.
+
+        Starts with the first circuit (index 0), then iterates over the other
+        circuits while folding them in. To fold two circuits together, they
+        are placed one after the other and then moved inward until any of:
+
+            a) Just before their operations would collide.
+            b) stop_at_first_alignment==True and their ends (or starts) align
+                for the first time.
+            c) Their ends (or starts) align for the last time.
+
+        Beware that this method is *not* associative. For example:
+
+            >>> a, b = cirq.LineQubit.range(2)
+            >>> A = cirq.Circuit(cirq.H(a))
+            >>> B = cirq.Circuit(cirq.H(b))
+            >>> f = cirq.Circuit.raggedy_add
+            >>> f(f(A, B), A) == f(A, f(B, A))
+            False
+            >>> len(f(f(f(A, B), A), B)) == len(f(f(A, f(B, A)), B))
+            False
+
+        Args:
+            circuits: The circuits to concatenate.
+            stop_at_first_alignment: Defaults to false. When true, the circuits
+                are never overlapped more than needed to align their starts or
+                to align their ends. When false, the smaller circuit can be
+                pushed deeper into the larger circuit, past the first time their
+                starts or ends align, until the second time their starts or ends
+                align.
+
+        Returns:
+            The concatenated and overlapped circuit.
+        """
+        if len(circuits) == 0:
+            return Circuit()
+        acc = Circuit(circuits[0])
+        for k in range(1, len(circuits)):
+            acc = _raggedy_add_helper(acc, circuits[k], stop_at_first_alignment)
+        return acc
+
+
+def _overlap_collision_time(
+    c1: 'cirq.AbstractCircuit', c2: 'cirq.AbstractCircuit', stop_at_first_alignment: bool
+) -> int:
+    # Tracks the first used moment index for each qubit in c2.
+    # Tracks the complementary last used moment index for each qubit in c1.
+    seen_times: Dict['cirq.Qid', int] = {}
+
+    # Start scanning from end of first and start of second.
+    upper_bound = (min if stop_at_first_alignment else max)(len(c1), len(c2))
```

Comment:
I was thinking this would be `len(c1)` if not stop_at_first_alignment, not the max of the two. Otherwise if c1 and c2 have zero overlaps and c2 is longer, it pushes c2 start to before c1.

Reply:
Pushing c2 to before c1 would correspond to the upper bound being the sum of the two lengths, not their max.

The bound has to be symmetric in the circuit lengths in order to correctly handle the case where c1 is much smaller and also the case where c2 is much smaller.

Reply:
An example here would be

```
c1 = cirq.Circuit(
    cirq.X(cirq.LineQubit(0)),
)
c2 = cirq.Circuit(
    cirq.H(cirq.LineQubit(1)),
    cirq.H(cirq.LineQubit(1)),
)
c3 = raggedy_add(c1, c2)
```

This results in the `X` from `c1` getting pushed forward to moment 1, which seems odd to me: I'd not expect `c1`'s positions to change. I'd expect `c2` never to get pushed back further than `c1`'s start, even if it's longer.
```
0: ───────X───
1: ───H───H───
```

Reply:
That's the intended behavior. If you want it to stop earlier, you set `stop_at_first_alignment=True`.

Reply:
The reasoning I'm using is that this operation should be symmetric in the sense that `f(a, b) == reverse(f(reverse(b), reverse(a))`.

Reply:
I could rename it to something besides `raggedy_add` to possibly make this clearer. I can see that `a.****_add(b)` prepending things into `a` could  be surprising.

Reply:
I see. I was thinking more the parameter would be something like `force_stop_when` and the values would be an enum either `rights_align` or `lefts_align`. 

One benefit of having `lefts_align` behavior would allow like if you have

circuit1a: k1 ops on q1
circuit1b: N-k1 ops on q1
circuit 2a: k2 ops on q2
circuit 2b: N-k2 ops on q2
circuit 3a: ...

and called `raggedy_add` on all of them, the resulting circuit would be a block of length N. I think the existing implementation would always end up staggered. IDK if that's a big deal or not.

Maybe call it `tetris_add`.

Reply:
Renamed to `tetris_concat`

Reply:
I like the new name

Reply:
@daxfohl - the block of length N behavior is what actually is happening in the current implementation with `stop_at_first_alignment=False`. 

```
from cirq import *

a,b,c = LineQubit.range(3)
c1 = Circuit(H(a), H(a), H(a))
c2 = Circuit(X(a), X(a))
c3 = Circuit(Y(b))
c4 = Circuit(Z(b),Z(b),Z(b),Z(b))

print(c1.tetris_concat(c2,c3,c4))
```

```
0: ───H───H───H───X───X───

1: ───Y───Z───Z───Z───Z───
```

Am I missing something? 

I like the name too! :)